app/models/docker_builder_service.rb

```diff
@@ -0,0 +1,93 @@
+require 'docker'
+
+class DockerBuilderService
+  attr_reader :build, :image
+
+  def initialize(build)
+    @build = build
+  end
+
+  def build!(image_name: nil, push: true)
+    # TODO: check if there's already a docker build
+
+    job = build.create_docker_build_job do |job|
+      job.project = build.project
+      job.user_id = build.created_by || NullUser.new.id
+      job.command = ''    # doesn't seem right, but command can't be nil
+      job.commit = build.git_sha
+      job.tag = build.git_ref
+    end
+
+    build.save!
+
+    job_execution = JobExecution.start_job(build.git_sha, job) do |execution, tmp_dir|
+      build_image(execution, tmp_dir, image_name, push)
+    end
+
+    job_execution.subscribe do
+      send_after_notifications
+    end
+  end
+
+  private
+
+  def build_image(execution, tmp_dir, image_name, push)
+    repository.executor = execution.executor
+    repository.setup!(tmp_dir, build.git_sha)
+
+    # TODO: figure out what to do about REVISION file
+    File.open("#{tmp_dir}/REVISION", 'w') { |f| f.write build.git_sha }
+
+    execution.output.write("Running Docker build")
+
+    @image = Docker::Image.build_from_dir(tmp_dir) do |output_chunk|
+      execution.output.write(parse_output_chunk(output_chunk))
+    end
+
+    build.docker_sha = @image.json['Id']
+    build.docker_ref = image_name || build.label || 'latest'
+    build.docker_image_url = "#{docker_registry_url}/#{build.project.docker_repo_name}/sha256:#{build.docker_sha}"
+    build.save!
+
+    @image.tag(repo: build.project.docker_repo_name, tag: build.docker_ref, force: true)
+
+    if push
+      # TODO: authenticate with the Docker registry
+      execution.output.write "Pushing Docker image to #{build.project.docker_repo_name}/#{build.docker_ref}"
+      @image.push do |output_chunk|
+        execution.output.write(parse_output_chunk(output_chunk))
+      end
+    end
+
+    @image
+  rescue Docker::Error::DockerError => e
```

Comment:
If I catch the error and return `nil`, then the job is marked as a failed.  If I allow the exception to bubble up, then it runs `JobExecution#error!`, which sets the job to be in an "error" state.  Maybe it's better to let the exception bubble up.